interop/test_utils.go

```diff
@@ -673,6 +676,89 @@ func DoPickFirstUnary(tc testgrpc.TestServiceClient) {
 	}
 }
 
+type soakIterationResult struct {
+	err     error
+	latency time.Duration
+}
+
+func doOneSoakIteration(tc testgrpc.TestServiceClient, resetChannel bool, serverAddr string, dopts []grpc.DialOption) error {
+	client := tc
+	if resetChannel {
+		conn, err := grpc.Dial(serverAddr, dopts...)
+		if err != nil {
+			return err
+		}
+		defer conn.Close()
+		client = testgrpc.NewTestServiceClient(conn)
+	}
+	// do a large-unary RPC
+	pl := ClientNewPayload(testpb.PayloadType_COMPRESSABLE, largeReqSize)
+	req := &testpb.SimpleRequest{
+		ResponseType: testpb.PayloadType_COMPRESSABLE,
+		ResponseSize: int32(largeRespSize),
+		Payload:      pl,
+	}
+	reply, err := client.UnaryCall(context.Background(), req)
+	if err != nil {
+		return fmt.Errorf("/TestService/UnaryCall RPC failed: %s", err)
+	}
+	t := reply.GetPayload().GetType()
+	s := len(reply.GetPayload().GetBody())
+	if t != testpb.PayloadType_COMPRESSABLE || s != largeRespSize {
+		return fmt.Errorf("got the reply with type %d len %d; want %d, %d", t, s, testpb.PayloadType_COMPRESSABLE, largeRespSize)
+	}
+	return nil
+}
+
+// DoSoakTest runs large unary RPCs in a loop for a configurable number of times, with configurable failure thresholds.
+// If resetChannel is false, then each RPC will be performed on tc. Otherwise, each RPC will be performed on a new
+// stub that is created with the provided server address and dial options.
+func DoSoakTest(tc testgrpc.TestServiceClient, serverAddr string, dopts []grpc.DialOption, resetChannel bool, soakIterations int, maxFailures int, perIterationMaxAcceptableLatency time.Duration, overallDeadline time.Time) {
+	start := time.Now()
+	var results []soakIterationResult
+	for i := 0; i < soakIterations; i++ {
+		if time.Now().After(overallDeadline) {
+			break
+		}
+		start := time.Now()
+		err := doOneSoakIteration(tc, resetChannel, serverAddr, dopts)
+		results = append(results, soakIterationResult{err: err, latency: time.Since(start)})
+	}
+	totalFailures := 0
+	hopts := stats.HistogramOptions{
+		NumBuckets:     20,
+		GrowthFactor:   1,
+		BaseBucketSize: 1,
+		MinValue:       0,
+	}
+	h := stats.NewHistogram(hopts)
+	for i := 0; i < len(results); i++ {
+		err := results[i].err
+		latencyMs := int64(results[i].latency / time.Millisecond)
+		if err != nil {
+			totalFailures++
+			fmt.Fprintf(os.Stderr, "soak iteration: %d elapsed_ms: %d failed: %s\n", i, latencyMs, err)
+		} else if results[i].latency > perIterationMaxAcceptableLatency {
+			totalFailures++
+			fmt.Fprintf(os.Stderr, "soak iteration: %d elapsed_ms: %d exceeds max acceptable latency: %d\n", i, latencyMs, perIterationMaxAcceptableLatency.Milliseconds())
+		} else {
+			fmt.Fprintf(os.Stderr, "soak iteration: %d elapsed_ms: %d succeeded\n", i, latencyMs)
+		}
+		h.Add(latencyMs)
```

Comment:
Nit: you can reorder this, and save those `else` statements

```go
h.add(latencyMs)
if err != nil {
  totalFailures++
  fmt.Fprintf()
  continue
}

if results[i].latency > perIterationMaxAcceptableLatency {
  totalFailures++
  fmt.Fprintf()
  continue
}
fmt.Fprintf(success)
```

Reply:
done